docs/proposals/scheduling/20230707-eviction-arbitrate-mechanism-in-descheduler.md

```diff
@@ -0,0 +1,229 @@
+
+## Motivation
+
+Arbitrate Mechanism is an important capability that Pod Migration relies on, and Pod Migration is relied on by many components (such as deschedulers). But Pod Migration is a complex process,  involving steps such as auditing, resource allocation, and application startup, and is mixed with application upgrading, scaling scenarios, and resource operation and maintenance operations by cluster administrators. 
+
+So when a large number of Pods are concurrently migrated, this may have some impact on the stability of the system. In addition, if many Pods of the same workload are migrated concurrently, it will also have an impact on the stability of the application. Moreover, If multiple jobs' Pods are migrated concurrently, it will cause a thundering herd effect. So we hope to handle the Pods in each job in sequence.
+
+Therefore, it is necessary to design an arbitration mechanism. This arbitration mechanism will select suitable `PodMigrationJob` to execute and control the execution speed of `PodMigrationJob` (to avoid a large number of jobs executing concurrently), thereby ensuring the stability of the system and application.
+
+### Goals
+
+1. Provide an Arbitrate Mechanism to sort and filter the PodMigrationJobs from arbitration queue periodically.
+2. Provide an event handler to intercepts event in creation status and place them in the arbitration queue.
+3. Defines an Arbitrate Mechanism Configuration, through which user can configure the arbitration mechanism.
+
+### Non-Goals/Future Work
+
+1. A new CRD or Controller.
+2. The generation process of PodMigrationJob.
+3. The specific execution process of PodMigrationJob.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+The descheduler evicts Pods with different priorities, and the importance of high priority Pods is also high. Therefore, The cost of migrating high priority Pods is greater than that of low priority Pods. So users expect low-priority Pods to be executed first to minimize migration costs.
+
+#### Story 2
+
+Multiple Pods of different workloads may be concurrently evicted by the descheduler. We hope that the migrated Pods can be dispersed across different workloads (like deployments) as much as possible, in order to avoid a decrease in service availability caused by the migration process.
+
+#### Story 3
+
+Multiple Pods in different jobs are evicted. We should migrate all Pods from one job at a time and process different jobs in sequence to avoid a thundering herd effect.
+
+### Implementation Details/Notes/Constraints
+
+#### Arbitration Mechanism
+
+The **PodMigrationJob Controller** will evaluate all PodMigrationJobs and select a batch of PodMigrationJobs from arbitration queue and add them to the reconcile work queue for executing. This selection process is called the **arbitration mechanism**. The arbitration mechanism includes three stages: `Sort`, `GroupFilter` and `Select`. And the reconcile work queue will help us to achieve rate limit.
```

Comment:
1. why not GrroupFilter First and Sort secondly.
2. If PMJ non-retrievable filter fails, current behaviors of controller is abort the job, this logic will remain the same?
3. PMJ retrievable filter currently acts as a rate-limiting of worklaod、namespace、node. What's the difference beween retrievable filter and GroupFilter.

Reply:
1. Because we need to remove low-priority PMJs during the GroupFilter, and Sort will help us to get PMJs' prioritys.
2. The GroupFilter here does not distinguish between non retrievable and retrievable. Failed PMJs will be arbitrated again and again until they succeed or deleted.
3. Retrievable filter is for a single PMJ, and GroupFilter is for a batch of PMJs. They are basically the same functionally, and we may use GroupFilter to replace Retrievable filter.

Reply:
/lgtm

Reply:
The arbitration mechanism replaces the existing filter mechanism in the original Reconcile, so compatibility with the original behavior must be considered. Therefore, it is necessary to consider calling a non-retryable Filter during arbitration, and if it fails, the PMJ should fail immediately.

Reply:
2. If a pmj created to migated a kube-system pod, pods of kube-system should be excluded, what is the expected behavior of PMJ arbitration.

Reply:
This PMJ will be set as very low priority in this batch of PMJs, and it will be passed to controller reconcile.

Reply:
> The arbitration mechanism replaces the existing filter mechanism in the original Reconcile, so compatibility with the original behavior must be considered. Therefore, it is necessary to consider calling a non-retryable Filter during arbitration, and if it fails, the PMJ should fail immediately.

It means that if some PMJs did not pass the filter they need to be immediately updated to failed phase and not passed to the controller reconcile?

Reply:
> > The arbitration mechanism replaces the existing filter mechanism in the original Reconcile, so compatibility with the original behavior must be considered. Therefore, it is necessary to consider calling a non-retryable Filter during arbitration, and if it fails, the PMJ should fail immediately.
> 
> It means that if some PMJs did not pass the filter they need to be immediately updated to failed phase and not passed to the controller reconcile?

Yes.

Reply:
In that case, I am going to add a step after GroupFilter to check each filtered job one by one if it should be tried again.